File: Claude/Simples/radixsort/radixsort.py
# Input: How to implement the radixsort sort algorithm? in python

def radix_sort(arr):
    """
    Performs radix sort on the given array.
    
    Args:
        arr (list): The input array to be sorted.
    
    Returns:
        list: The sorted array.
    """
    # Get the maximum element to know the number of digits
    max_element = max(arr)

    # Do counting sort for every digit. Note that instead of passing
    # digit number, exp is passed. exp is 10^i where i is current
    # digit number
    exp = 1
    while max_element / exp > 0:
        counting_sort(arr, len(arr), exp)
        exp *= 10

    return arr

def counting_sort(arr, n, exp):
    """
    Performs counting sort on the given array based on the current digit.
    
    Args:
        arr (list): The input array to be sorted.
        n (int): The length of the input array.
        exp (int): The current digit to sort on (10^i).
    
    Returns:
        None (the input array is modified in-place)
    """
    output = [0] * n  # output array
    count = [0] * 10  # count array as 0-9

    # Store count of occurrences in count[]
    for i in range(n):
        index = arr[i] // exp
        count[index % 10] += 1

    # Change count[i] so that count[i] now contains actual
    # position of this digit in output array
    for i in range(1, 10):
        count[i] += count[i - 1]

    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1
        i -= 1

    for i in range(n):
        arr[i] = output[i]

example_array = [170, 45, 75, 90, 802, 24, 2, 66]
sorted_array = radix_sort(example_array)
print(sorted_array)  # Output: [2, 24, 45, 66, 75, 90, 170, 802]

"""
Score: 1
"""

File: Claude/Simples/radixsort/radixsort.cpp
// Input: How to implement the radixsort sort algorithm? in c++

#include <iostream>
#include <vector>

// Function to get the maximum element from the array
int getMax(std::vector<int>& arr, int n) {
    int mx = arr[0];
    for (int i = 1; i < n; i++)
        if (arr[i] > mx)
            mx = arr[i];
    return mx;
}

// Counting sort based on the digit. exp is 10^i where i is current digit level.
void countSort(std::vector<int>& arr, int n, int exp) {
    std::vector<int> output(n);
    int count[10] = { 0 };

    // Store count of occurrences in count[]
    for (int i = 0; i < n; i++)
        count[(arr[i] / exp) % 10]++;

    // Change count[i] so that count[i] now contains actual
    // position of this digit in output array
    for (int i = 1; i < 10; i++)
        count[i] += count[i - 1];

    // Build the output array
    int i = n - 1;
    while (i >= 0) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
        i--;
    }

    // Placing output array elements back to arr[]
    for (i = 0; i < n; i++)
        arr[i] = output[i];
}

// The main function to implement radix sort
void radixSort(std::vector<int>& arr, int n) {
    // Find the maximum element to know the number of digits
    int m = getMax(arr, n);

    // Do counting sort for every digit. Note that instead of passing digit number,
    // exp is passed. exp is 10^i where i is current digit number
    for (int exp = 1; m / exp > 0; exp *= 10)
        countSort(arr, n, exp);
}

int main() {
    std::vector<int> arr = { 170, 45, 75, 90, 802, 24, 2, 66 };
    int n = arr.size();

    radixSort(arr, n);

    std::cout << "Sorted array: ";
    for (int i = 0; i < n; i++)
        std::cout << arr[i] << " ";
    std::cout << std::endl;

    return 0;
}

/*
Score: 1
*/